src/aggregation/bucket/histogram/histogram.rs

```diff
@@ -0,0 +1,822 @@
+    pub min_doc_count: Option<u64>,
+    /// Sets a hard limit for the data range.
+    /// This can be used to filter values if they are not in the data range.
+    pub hard_bounds: Option<HistogramBounds>,
+    /// Can be set to extend your bounds. The range of the buckets is by default defined by the
+    /// data range of the values of the documents. As the name suggests, this can only be used to
+    /// extend the value range. If the bounds for min or max are not extending the range, the value
+    /// has no effect on the returned buckets.
+    ///
+    /// Cannot be set in conjunction with min_doc_count > 0, since the empty buckets from extended
+    /// bounds would not be returned.
+    pub extended_bounds: Option<HistogramBounds>,
+}
+
+impl HistogramAggregation {
+    fn validate(&self) -> crate::Result<()> {
+        if self.interval <= 0.0f64 {
+            return Err(TantivyError::InvalidArgument(
+                "interval must be a positive value".to_string(),
+            ));
+        }
+
+        if self.min_doc_count.unwrap_or(0) > 0 && self.extended_bounds.is_some() {
+            return Err(TantivyError::InvalidArgument(
+                "Cannot set min_doc_count and extended_bounds at the same time".to_string(),
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Returns the minimum number of documents required for a bucket to be returned.
+    pub fn min_doc_count(&self) -> u64 {
+        self.min_doc_count.unwrap_or(0)
+    }
+}
+
+/// Used to set extended or hard bounds on the histogram.
+#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
+pub struct HistogramBounds {
+    /// The lower bounds.
+    pub min: f64,
+    /// The upper bounds.
+    pub max: f64,
+}
+
+impl HistogramBounds {
+    fn contains(&self, val: f64) -> bool {
+        val >= self.min && val <= self.max
+    }
+}
+
+/// The collector puts values from the fast field into the correct buckets and does a conversion to
+/// the correct datatype.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentHistogramCollector {
+    /// The buckets containing the aggregation data.
+    buckets: Vec<SegmentHistogramBucketEntry>,
+    field_type: Type,
+    req: HistogramAggregation,
+    offset: f64,
+    first_bucket_num: i64,
+    bounds: HistogramBounds,
+}
+
+impl SegmentHistogramCollector {
+    pub fn into_intermediate_bucket_result(self) -> IntermediateBucketResult {
+        // We cut off the empty buckets at the start and end to mimic elasticsearch
+        // behaviour
+        let skip_start = self
+            .buckets
+            .iter()
+            .take_while(|bucket| bucket.doc_count == 0)
+            .count();
+        let skip_end = self
+            .buckets
+            .iter()
+            .rev()
+            .take_while(|bucket| bucket.doc_count == 0)
+            .count();
+        let num_buckets = self.buckets.len();
+
+        let buckets = self
```

Comment:
I understand the saturating_sub tries to address the edge case, but that's a cryptic way to write code.

Reply:
since you solved the riddle I can remove it ;)
I just realized we can simply filter all empty buckets.